File: src/bin/0322-coin-change.cpp
// @leetup=info id=322 lang=cpp slug=coin-change

#include <algorithm>
#include <cassert>
#include <climits>
#include <iostream>
#include <map>
#include <set>
#include <tuple>
#include <vector>

// @leetup=code

class Solution {
 public:
  int coinChange(std::vector<int>& coins, int amount) {
    std::vector<size_t> dp(amount + 1, INT_MAX);
    dp[0] = 0;
    for (int i = 1; i <= amount; ++i) {
      for (int coin : coins) {
        if (i - coin >= 0) {
          dp[i] = std::min(dp[i], dp[i - coin] + 1);
        }
      }
    }
    return dp[amount] == INT_MAX ? -1 : dp[amount];
  }
};
// @leetup=code

int main() {
  std::cout << "Hello, world!" << std::endl;
  return 0;
}

File: src/bin/0322-coin-change.py
# @leetup=info id=322 lang=python3 slug=coin-change



# @leetup=code


class Solution:
    def coinChange(self, coins: list[int], amount: int) -> int:
        dp = [inf] * (amount + 1)
        dp[0] = 0

        for a in range(1, amount + 1):
            for coin in coins:
                # if a - coin >= 0 and dp[a-coin] != inf:
                if a - coin >= 0:
                    dp[a] = min(dp[a], dp[a - coin] + 1)

        return dp[amount] if dp[amount] != inf else -1


# @leetup=code
